assets/content/cookbook/Functional Programming/enum-gadt.md

```diff
@@ -0,0 +1,87 @@
+[tags]: / "enum,pattern-matching,functional-programming"
+
+# Enums as GADTs
+
+As [already established](http://code.haxe.org/category/beginner/enum-adt.html) Haxe enums are a form of algebraic data types. In fact, they may even serve as so called "*generalized* algebraic data types" - GADTs for short. While for an "ordinary" enum every constructor yields the same type, with an GADT each constructor may yield a different type. 
+
+To illustrate that, let's define a little language for arithmetic expressions:
+
+```haxe
+enum Expr<T> {
+
+  Sum(a:Expr<Float>, b:Expr<Float>):Expr<Float>; 
+  Product(a:Expr<Float>, b:Expr<Float>):Expr<Float>;
+  Power(a:Expr<Float>, b:Expr<Float>):Expr<Float>; // <-- this constructor returns an Expr<Float> ...
+  
+  GreaterThan(a:Expr<Float>, b:Expr<Float>):Expr<Bool>;// <-- ... and this one an Expr<Bool>
+  Not(a:Expr<Bool>):Expr<Bool>;
+  Or(a:Expr<Bool>, b:Expr<Bool>):Expr<Bool>;
+  And(a:Expr<Bool>, b:Expr<Bool>):Expr<Bool>;
+  
+  Const(v:T):Expr<T>;
+  Equals<O>(a:Expr<O>, b:Expr<O>):Expr<Bool>;
+}
+```
+
+So now we can say: I want a numeric expression or a boolean one, by either saying `Expr<Float>` or `Expr<Bool>`.
+
+The last two constructors in the example are particularly interesting:
+
+- `Const` may accept a value of any type and becomes and `Expr` of that type. 
+- `Equals` has a type parameter. This is actually not GADT specific. Ordinary enum constructors may have this too, because at the bottom line they are functions and may therefore be parametrized. In the case of `Equals` it is the type of the `Expr` being compared. It is arbitrary, but still must be equal for both operands and the result will always be boolean. This models very closely how `==` works.
+
+For example `1 + 1 == 2` could be written as `Equals(Sum(Const(1) + Const(1)), Const(2))` and will compile, as opposed to `Equals(Const(3.14), Const('test'))` which will fail with `String should be Float` exactly as `3.14 == 'test'`. 
```

Comment:
`Equals(Sum(Const(1) + Const(1)), Const(2))` should be `Equals(Sum(Const(1), Const(1)), Const(2))` but that [doesn't compile](http://try.haxe.org/#D1bF2) either due (in part) to `Sum` expecting arguments of type `Expr<Float>`.